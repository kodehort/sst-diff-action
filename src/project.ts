import fsSync from 'fs'
import path from 'path'
import core from '@actions/core'

const CONFIG_EXTENSIONS = [
  '.config.ts',
  '.config.mts',
  '.config.cts',
  '.config.cjs',
  '.config.mjs',
  '.config.js',
]

interface Options {
  stage: string
  root?: string
}

export interface Project {
  config: { stage: string }
  paths: {
    config: string
    out: string
    dist: string
  }
}

export async function initProject(globals: Options) {
  core.debug('initing project')
  const root = globals.root || (await findRoot())
  const out = path.join(root, '.sst')
  let file: string | undefined
  for (const ext of CONFIG_EXTENSIONS) {
    file = path.join(root, `sst${ext}`)
    if (!fsSync.existsSync(file)) continue
    core.debug('found sst config')
  }

  if (!file) {
    throw new Error('Could not find a config file')
  }

  const stage = globals.stage
  const project: Project = {
    config: {
      stage,
    },
    paths: {
      config: file,
      out,
      dist: path.join(out, 'dist'),
    },
  }

  core.debug('Config loaded')
  core.debug(JSON.stringify(project))

  return project
}

async function findRoot() {
  async function find(dir: string): Promise<string> {
    if (dir === '/') throw new Error('Could not find a config file')

    for (const ext of CONFIG_EXTENSIONS) {
      const configPath = path.join(dir, `sst${ext}`)
      if (fsSync.existsSync(configPath)) {
        return dir
      }
    }
    return await find(path.join(dir, '..'))
  }
  const result = await find(process.cwd())
  return result
}
