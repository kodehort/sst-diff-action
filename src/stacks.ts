import { ArtifactMetadataEntryType } from '@aws-cdk/cloud-assembly-schema'
import type { FormatStream } from '@aws-cdk/cloudformation-diff'
import { diffTemplate, formatDifferences } from '@aws-cdk/cloudformation-diff'
import { type Template } from 'aws-cdk-lib/assertions'
import type { CloudFormationStackArtifact } from 'aws-cdk-lib/cx-api'
import { CloudAssembly } from 'aws-cdk-lib/cx-api'

export * as Stacks from './stacks'

export async function diff(
  stack: CloudFormationStackArtifact,
  oldTemplate: Template,
) {
  // Generate diff
  const diffOutput = diffTemplate(oldTemplate, stack.template)
  if (diffOutput.isEmpty) {
    return { count: 0 }
  }

  // Only display resource and output changes
  // @ts-ignore
  diffOutput.iamChanges = { hasChanges: false }
  // @ts-ignore
  diffOutput.securityGroupChanges = { hasChanges: false }
  // @ts-ignore
  diffOutput.awsTemplateFormatVersion = false
  // @ts-ignore
  diffOutput.transform = false
  // @ts-ignore
  diffOutput.description = false
  // @ts-ignore
  diffOutput.parameters = { differenceCount: 0 }
  // @ts-ignore
  diffOutput.metadata = { differenceCount: 0 }
  // @ts-ignore
  diffOutput.mappings = { differenceCount: 0 }
  // @ts-ignore
  diffOutput.conditions = { differenceCount: 0 }
  // @ts-ignore
  diffOutput.unknown = { differenceCount: 0 }

  // Filter out SST internal diffs
  // @ts-ignore
  delete diffOutput.outputs.diffs?.['SSTMetadata']

  // Format diff
  const output: string[] = []
  const stream = {
    write(chunk: string) {
      output.push(`   ${chunk}`)
    },
  } as FormatStream
  const pathMap = await buildLogicalToPathMap(stack)
  formatDifferences(stream, diffOutput, pathMap)

  // Remove trailing newline
  while (true) {
    if (output[output.length - 1]?.match(/^\s*$/)) {
      output.pop()
    } else {
      break
    }
  }

  return {
    count:
      diffOutput.outputs.differenceCount + diffOutput.resources.differenceCount,
    diff: output.join(''),
  }
}

async function buildLogicalToPathMap(stack: CloudFormationStackArtifact) {
  const map: { [id: string]: string } = {}
  for (const md of stack.findMetadataByType(
    ArtifactMetadataEntryType.LOGICAL_ID,
  )) {
    map[md.data as string] = md.path
  }
  return map
}

export async function loadAssembly(from: string) {
  return new CloudAssembly(from)
}
