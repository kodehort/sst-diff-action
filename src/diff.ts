import { summary } from '@actions/core'
import {
  CloudFormationClient,
  GetTemplateCommand,
} from '@aws-sdk/client-cloudformation'
import { type Template } from 'aws-cdk-lib/assertions'

import { type Project } from './project'
import { Stacks } from './stacks'

const stackNameToId = (stackName: string) => {
  return stackName
}

export const writeSummary = async (project: Project): Promise<void> => {
  // Build app
  const assembly = Stacks.loadAssembly(project.paths.dist)

  summary.addHeading(
    `Changes that will be deployed to ${project.config.stage}`,
    2,
  )

  // Diff each stack
  let changesAcc = 0
  let changedStacks = 0
  for (const stack of assembly.stacks) {
    // get old template
    const oldTemplate = await getTemplate(stack.stackName)
    if (!oldTemplate) {
      summary.addHeading(`${stackNameToId(stack.stackName)}: New stack`, 3)
      continue
    }

    // generate diff
    const { count, diff } = Stacks.diff(stack, oldTemplate)

    // print diff result
    if (count === 0) {
      summary.addHeading(`${stackNameToId(stack.stackName)}: No changes`, 3)
    } else if (count === 1) {
      summary.addHeading(
        `${stackNameToId(stack.stackName)}: ${count} change`,
        3,
      )
      if (diff) {
        summary.addCodeBlock(diff)
      }

      changesAcc += count
      changedStacks++
    } else {
      summary.addHeading(
        `${stackNameToId(stack.stackName)}: ${count} changes`,
        3,
      )
      if (diff) {
        summary.addCodeBlock(diff, 'diff')
      }

      changesAcc += count
      changedStacks++
    }

    summary.addBreak()
  }

  summary.addSeparator()
  // Handle no changes
  if (changedStacks === 0) {
    summary.addRaw('No changes')
  } else {
    summary.addRaw(
      `${changesAcc === 1 ? '1 change found in' : `${changesAcc} changes in`} ${
        changedStacks === 1 ? '1 stack' : `${changedStacks} stacks`
      } to be deployed.`,
    )
  }

  await summary.write()
}

async function getTemplate(stackName: string): Promise<Template | undefined> {
  const client = new CloudFormationClient({})
  const response = await client.send(
    new GetTemplateCommand({ StackName: stackName }),
  )
  if (!response.TemplateBody) return undefined
  return JSON.parse(response.TemplateBody) as Template
}
