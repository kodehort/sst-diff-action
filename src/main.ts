import core from '@actions/core'
import { writeSummary } from './diff'
import { initProject } from './project'

async function run(): Promise<void> {
  try {
    const stage: string = core.getInput('stage')
    // debug is only output if you set the secret `ACTIONS_STEP_DEBUG` to true
    core.debug(`Running diff against stage ${stage}`)
    const project = await initProject({ stage })
    await writeSummary(project)
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}

run()
