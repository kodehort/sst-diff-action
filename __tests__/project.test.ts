import * as fs from 'fs'
import path from 'path'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { initProject } from '../src/project'

const testDirectoryPath = path.join(__dirname, 'test')
const testConfigFilePath = path.join(testDirectoryPath, 'sst.config.ts')
const testDistDirectoryPath = path.join(testDirectoryPath, '.sst', 'dist')

describe('project', () => {
  beforeEach(async () => {
    await fs.promises.mkdir(testDirectoryPath, { recursive: true })
    await fs.promises.mkdir(testDistDirectoryPath, { recursive: true })
    await fs.promises.writeFile(testConfigFilePath, '', { encoding: 'utf8' })
  })

  afterEach(async () => {
    await fs.promises.unlink(testConfigFilePath)
  })

  it('throws an error if no config file is found', async () => {
    await expect(initProject({ stage: 'test' })).rejects.toThrow(
      'Could not find a config file',
    )
  })

  it('locates the sst project as expected', async () => {
    vi.spyOn(process, 'cwd').mockReturnValue(testDirectoryPath)
    const project = await initProject({ stage: 'test' })
    expect(project).toMatchSnapshot()
  })
})
