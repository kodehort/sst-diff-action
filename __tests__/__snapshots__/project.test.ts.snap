// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`project > locates the sst project as expected 1`] = `
{
  "config": {
    "stage": "test",
  },
  "paths": {
    "config": "/Users/alistairstead/code/sst-diff-action/__tests__/test/sst.config.js",
    "dist": "/Users/alistairstead/code/sst-diff-action/__tests__/test/.sst/dist",
    "out": "/Users/alistairstead/code/sst-diff-action/__tests__/test/.sst",
  },
}
`;
